Introduce `tokio-trace`

hawkw opened the pull request:
<!-- Thank you for your Pull Request. Please provide a description above
and review the requirements below.

Bug fixes and new features should include tests.

Contributors guide:
https://github.com/tokio-rs/tokio/blob/master/CONTRIBUTING.md -->

## Motivation

In asynchronous systems like Tokio, interpreting traditional log
messages can often be quite challenging. Since individual tasks are
multiplexed on the same thread, associated events and log lines are
intermixed making it difficult to trace the logic flow. Currently, none
of the available logging frameworks or libraries in Rust offer the
ability to trace logical paths through a futures-based program.

There also are complementary goals that can be accomplished with such a
system. For example, metrics / instrumentation can be tracked by
observing emitted events, or trace data can be exported to a distributed
tracing or event processing system.

In addition, it can often be useful to generate this diagnostic data in
a structured manner that can be consumed programmatically. While prior
art for structured logging in Rust exists, it is not currently
standardized, and is not "Tokio-friendly".

## Solution

This branch adds a new library to the tokio project, `tokio-trace`.
`tokio-trace` expands upon logging-style diagnostics by allowing
libraries and applications to record structured events with additional
information about *temporality* and *causality* --- unlike a log
message, a span in `tokio-trace` has a beginning and end time, may be
entered and exited by the flow of execution, and may exist within a
nested tree of similar spans. In addition, `tokio-trace` spans are
*structured*, with the ability to record typed data as well as textual
messages.

The `tokio-trace-core` crate contains the core primitives for this
system, which are expected to remain stable, while `tokio-trace` crate
provides a more "batteries-included" API. In particular, it provides
macros which are a superset of the `log` crate's `error!`, `warn!`,
`info!`, `debug!`, and `trace!` macros, allowing users to begin the
process of adopting `tokio-trace` by performing a drop-in replacement.

## Notes

Work on this project had previously been carried out in the
[tokio-trace-prototype] repository. In addition to the `tokio-trace` and
`tokio-trace-core` crates, the `tokio-trace-prototype` repo also
contains prototypes or sketches of adapter, compatibility, and utility
crates which provide useful functionality for `tokio-trace`, but these
crates are not yet ready for a release. When this branch is merged, that
repository will be archived, and the remaining unstable crates will be
moved to a new `tokio-trace-nursery` repository. Remaining issues on the
`tokio-trace-prototype` repo will be moved to the appropriate new repo.

The crates added in this branch are not _identical_ to the current head
of the `tokio-trace-prototype` repo, as I did some final clean-up and docs
polish in this branch prior to opening this PR.

[tokio-trace-prototype]: https://github.com/hawkw/tokio-trace-prototype

Refs: #561

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

davidbarsky commented on tokio-trace/README.md:
````diff
@@ -0,0 +1,156 @@
+extern crate tokio_trace;
+```
+
+This crate provides macros for creating `Span`s and `Event`s, which represent
````
A useful thing to expand upon is when someone should use a `span` versus an `event`, and at what levels of granularity. A starting point would be something like:

- Use spans for things that correspond to a discrete unit of work (i.e., a given request's lifespan in a server with a quick request/response cycle) or use a span for each IO-action/each time you interact with a new external system (a database, a queue, whatever).
- Use events for things that happened _within_ a span—a request returned with a given status code, grabbed *n* items from a queue, etc.

hawkw replied:
:+1: this is good advice, thanks --- do you think this kind of discussion is more appropriate for the README or for RustDoc?

davidbarsky replied:
Both locations, I think, opting for greater detail in the Rustdoc.

hawkw replied:
Added to the docs in 355a3d9 :)

davidbarsky commented on tokio-trace/README.md:
````diff
@@ -0,0 +1,156 @@
+}
+```
+
+You can find examples showing how to use this crate in the examples directory.
````
Should the examples include a hyper server being instrumented?

hawkw replied:
I've written such an example, but it relies on crates in the tokio-trace ecosystem that aren't yet stable. In practice, since `tokio-trace` itself doesn't provide any subscriber implementations that produce output, most non-trivial examples will depend on crates that aren't stable *yet*.

davidbarsky replied:
Ah, yes, I remember those. I forgot stability would be a concern.

hawkw replied:
Ideally, an instrumented Hyper server would also probably want to use the `futures` compatibility crate, which is also not yet stabilized. I agree that we ought to eventually have an example like that here, but I don't think it can be added in _this_ branch. We'll add it in the future when the compatibility crates are more stable.

seanmonstar commented on tokio-trace/LICENSE:
```diff
@@ -0,0 +1,25 @@
+Copyright (c) 2018 Tokio Contributors
```
```suggestion
Copyright (c) 2019 Tokio Contributors
```

😉

seanmonstar commented on tokio-trace/src/dispatcher.rs:
```diff
@@ -0,0 +1,37 @@
+/// [`Event`]: ::Event
+pub fn with_default<T>(dispatcher: Dispatch, f: impl FnOnce() -> T) -> T {
+    let prior = CURRENT_DISPATCH.try_with(|current| current.replace(dispatcher));
+    let result = f();
```
Should this use some drop guard so that the prior dispatcher is set back even if the closure panics?

hawkw replied:
Switched to use a drop guard in 9c38d67074d2d24e928c4fc52ddae8186904edf9. Good call!

seanmonstar commented on tokio-trace/src/field.rs:
```diff
@@ -0,0 +1,225 @@
+/// Indexing a field with a string results in an iterative search that performs
+/// string comparisons. Thus, if possible, once the key for a field is known, it
+/// should be used whenever possible.
+pub trait AsField {
```
Is it expected that people would implement this trait externally, or should it be sealed?

seanmonstar commented on tokio-trace/src/field.rs:
```diff
@@ -0,0 +1,225 @@
+    fn as_field(&self, metadata: &Metadata) -> Option<Field>;
+}
+
+pub trait Record {
```
Is it expected that people would implement this trait externally, or should it be sealed?

seanmonstar commented on tokio-trace/src/field.rs:
```diff
@@ -0,0 +1,225 @@
+/// Implementors of `Value` may call the appropriate typed recording methods on
+/// the `Subscriber` passed to `record` in order to indicate how their data
+/// should be recorded.
+pub trait Value {
```
Is it expected that people would implement this trait externally, or should it be sealed?

hawkw replied:
`Value` is intended to be user-implemented. In the future, when the value-recording API is more fully featured (i.e., it supports serializing structured values such as lists, key-value pairs etc, as prototyped [here](https://github.com/hawkw/tokio-trace-prototype/pull/93)), I expect most user-defined types to eventually be recorded using their `Value` implementations rather than `field::display`/`field::debug`.

seanmonstar commented on tokio-trace/src/field.rs:
```diff
@@ -0,0 +1,225 @@
+
+/// Wraps a type implementing `fmt::Display` as a `Value` that can be
+/// recorded using its `Display` implementation.
+pub fn display<'a, T>(t: T) -> DisplayValue<T>
```
Does this `'a` lifetime do anything? Doesn't appear to...

hawkw replied:
Whoops, that's left over from earlier code; removed it in e16bfd0.

seanmonstar commented on tokio-trace/src/lib.rs:
````diff
@@ -0,0 +1,760 @@
+//! Next, add this to your crate:
+//!
+//! ```rust
+//! #[macro_use]
````
Since this is a new crate, should we just start showing how to import with the newer macro importing ability? AKA, `use tokio_trace::span`, etc?

hawkw replied:
The newer macro import syntax doesn't work on Rust 1.26.0, which is tokio's minimum supported Rust version. We could use the new syntax, but then we'd have to stop trying to compile this example as a doctest, as it would break the 1.26.0 CI build.

seanmonstar commented on tokio-trace/src/lib.rs:
```diff
@@ -0,0 +1,760 @@
+        event!(target: $target, $crate::Level::TRACE, { $($k $( = $val)* ),* })
+    );
+    (target: $target:expr, $($arg:tt)+ ) => (
+        drop(event!(target: $target, $crate::Level::TRACE, {}, $($arg)+));
```
Why does this drop the event, but the others don't? If there's good reason, perhaps a tiny comment for Future Us?

hawkw replied:
Added a comment in c37623000e53dfdf5f4d3e4aa832b4da12f6a2cf.

carllerche commented on tokio-trace/src/lib.rs:
```diff
@@ -0,0 +1,774 @@
+    field::Value,
+    span::{Event, Id, Span},
+    subscriber::Subscriber,
+    tokio_trace_core::{
```
Most of these probably should be `doc(hidden)`

hawkw replied:
Done in 9478dbe1cc337b858b68814754206fc7d3393fd9.

hawkw commented on tokio-trace/tokio-trace-core/src/lib.rs:
```diff
@@ -0,0 +1,114 @@
+/// [`Callsite`]: ::callsite::Callsite
+/// [`Callsite::id`]: ::callsite::Callsite::id
+#[macro_export]
+macro_rules! identify_callsite {
```
This is essentially just a stand-in for a `const fn` constructor for `callsite::Identifier`, so that it can be constructed publically even though the inner field is `#[doc(hidden)]` and it appears to be private in the docs. I'll add an example. 👍

hawkw commented on tokio-trace/tokio-trace-core/src/subscriber.rs:
```diff
@@ -0,0 +1,353 @@
+    ///
+    /// If all active subscribers are `NEVER` interested in a callsite, it will
+    /// be completely disabled unless a new subscriber becomes active.
+    pub const NEVER: Interest = Interest(InterestKind::Never);
```
Thanks for bringing up the potential forward-compatibility hazard of using consts, I hadn't thought of that. I think these weren't functions because I wanted to avoid having to potentially push a stack frame to construct a new interest, but they can probably be changed.

hawkw commented on tokio-trace/src/span.rs:
```diff
@@ -0,0 +1,692 @@
+/// span will silently do nothing. Thus, the handle can be used in the same
+/// manner regardless of whether or not the trace is currently being collected.
+#[derive(Clone, PartialEq, Hash)]
+pub struct Span<'a> {
```
I had wondered about that when the lifetime was added (https://github.com/hawkw/tokio-trace-prototype/pull/157#issuecomment-445084549), too, and I'm still not sold on it.

In the `tokio-trace-futures` lib (in `tokio-trace-nursery`), IIRC, we just have hardcoded everything as `Span<'static>`.  I thought it might be better to be more flexible here, and have the `-futures` crate introduce the static requirement. But, it does add more noise. I'd be happy to change this to always be `'static`.

hawkw commented on tokio-trace/src/span.rs:
```diff
@@ -0,0 +1,692 @@
+
+    /// Signals that this span should close the next time it is exited, or when
+    /// it is dropped.
+    pub fn close(&mut self) {
```
Span closure _is_ tracked on drop; this function simply drops the inner span _now_ (calling `Subscriber::drop_id`), and leaving an empty disabled handle. This is intended for cases where the handle is a member of some struct with a longer lifetime, but the span should close.

I think the comment may need to be updated here.

hawkw replied:
Comment updated in 5acc8bc014e874d4893cf9df802b634eebe3cd40

hawkw commented on tokio-trace/src/dispatcher.rs:
```diff
@@ -0,0 +1,46 @@
+    f()
+}
+
+pub(crate) fn with_current<T, F>(mut f: F) -> T
```
Yeah, I find the naming around this really disappointing but I haven't been able to come up with anything better.

hawkw commented on tokio-trace/src/lib.rs:
```diff
@@ -0,0 +1,774 @@
+
+/// Constructs a new static callsite for a span or event.
+#[macro_export]
+macro_rules! callsite {
```
I think I misunderstood https://github.com/hawkw/tokio-trace-prototype/issues/150 as implying that this _should_ be public API, but it looks like that wasn't actually the intention behind that issue. I'll hide this.

hawkw replied:
Hidden in 9478dbe1cc337b858b68814754206fc7d3393fd9.

hawkw commented on tokio-trace/src/span.rs:
```diff
@@ -0,0 +1,692 @@
+/// enabled by the current filter. This type is primarily used for implementing
+/// span handles; users should typically not need to interact with it directly.
+#[derive(Debug)]
+pub(crate) struct Enter<'a> {
```
This began its life as an `Inner`, but I thought `Enter` was more descriptive as it described what the type was used _for_. I'm happy to change it back to `Inner` (or to something else, if anyone has a better suggestion?)

hawkw replied:
Renamed to `Inner` in 37c4dcb943165ad04b6109b40badd9a3bc4748e1.

hawkw commented on tokio-trace/src/lib.rs:
```diff
@@ -0,0 +1,774 @@
+// Somehow this `use` statement is necessary for us to re-export the `core`
+// macros on Rust 1.26.0. I'm not sure how this makes it work, but it does.
+#[allow(unused_imports)]
+use tokio_trace_core::*;
```
I think `metadata!` should be in the docs, but if you disagree we can hide it for sure.

hawkw replied:
Hidden in 9478dbe1cc337b858b68814754206fc7d3393fd9.

hawkw commented on tokio-trace/src/subscriber.rs:
```diff
@@ -0,0 +1,63 @@
+///
+/// This is intended for use by `Subscriber` implementations.
+#[derive(Clone)]
+pub struct CurrentSpanPerThread {
```
Removed in 39eded26e51bbaa23c899ec543cb28d5a90e2c4d, will eventually have this in `-subscriber`.

hawkw commented:
@carllerche Thanks for the thorough review! I believe I've addressed all of your inline comments (mostly by making the requested changes).

Regarding your other  comments:
> * Should there be a `debug_span!` macro and a `DebugSpan` type? When compiled w/o debug_assertions, it would be a unit struct.

I like this idea (although we might also want to consider the `debug_span!` name for a macro that constructs a span at the `DEBUG` verbosity level, I wish that name was less overloaded). Since this sounds like a purely additive change, do you think it ought to be made in an additional PR after this one lands?

> * The public API of `tokio-trace` should probably be pruned down to focus on instrumentation (not subscribers). Any API that isn't needed should be removed or marked as `#[doc(hidden)]`.

I _think_ I've addressed this by making most of the changes you requested --- is there anything else that's currently re-exported that you'd like to see removed?

davidbarsky commented:
Docs nit: can you do a find and replace of `---` for `—`? The em dashes aren't being rendered by rustdoc.

stjepang commented:
A few comments after reading the docs. Pardon me if some of the questions are silly. :)

* `Span::enter()` takes a `&mut self`. When does it make sense to enter a single span multiple times?

* If I need to trace something in a loop, should I create a span outside it and call `span.enter(...)` inside the loop, or invoke `span!` in every iteration as in `span!(...).enter(...)`?

* Why do we need custom macro `event!(Level::INFO, ...)`? Would it be possible to hook `tokio-trace` into the `log` crate so that `log::info!(...)` does the exact same thing? (asking as someone not super familiar with logging in Rust)

* In executable, we need `subscriber::with_default()` to set up the subscriber. Would it make sense for the Tokio runtime to do this automatically?

* The docs gave me a rough idea of what `tokio-trace` does. It makes sense why this is more useful than old fashioned logging, but it'd be nice to see a "hello world"/"quickstart" example with the result of `tokio-trace`. In other words, docs show us how the crate is used, but doesn't show the *result* it produces.

* Are there any pitfalls one should be aware of? For example, I believe spans and events must be created inside the Tokio runtime for them to work - is this correct? AFAIK, this is a big reason why we're introducing [`tokio::start()`](https://github.com/tokio-rs/tokio/issues/840).

KodrAus commented on tokio-trace/src/lib.rs:
```diff
@@ -0,0 +1,780 @@
+//! unstructured logging and span-based tracing. Similar to log records, they
+//! may be recorded at a number of levels, and can have unstructured,
+//! human-readable messages; however, they also carry key-value data and exist
+//! within the context of the tree of spans that comprise a trase. Thus,
```
nit: trase

KodrAus commented on tokio-trace/src/lib.rs:
```diff
@@ -0,0 +1,780 @@
+                    $(
+                        let key = keys.next()
+                            .expect(concat!("metadata should define a key for '", stringify!($k), "'"));
+                        event!(@ record: event, $k, &key, $($val)*);
```
I think this could be tricky for frameworks like `slog` (and maybe also `log`) that expect the structured data to be wholly and readily accessible rather than streamed one value at a time. While values are either `Debug` or one of a few primitive types this probably isn't a problem to work around, but might be something we should think about as the value API evolves?

In my opinion, having to box values for integrating with other frameworks is ok.

hawkw replied:
Thanks for your feedback --- compatibility with other frameworks here is definitely an important consideration.

> I think this could be tricky for frameworks like slog (and maybe also log) that expect the structured data to be wholly and readily accessible rather than streamed one value at a time.

I think an adapter for a framework like `slog` could solve this problem by waiting for all the field values to be added prior to exposing an event to `slog` --- this is part of the motivation behind why `tokio-trace` stipulates that all the fields on an event must be _declared_ when the event is created, even if their values are not provided at that time. This way,  we know when the event's fields are "complete". The adapter would have to serialize/format all the field values individually as they come in before exposing them to the other framework. I don't _think_ that's an issue but I could be incorrect.

> In my opinion, having to box values for integrating with other frameworks is ok.

I'm also fine with having integration code/adapters allocate. However, @carllerche has stressed that ensuring the instrumentation code should never allocate per-field or per-event is a major design goal for `tokio-trace`. We should ensure that the choice of whether or not to allocate here is left to the _consumer_ of the trace data.

KodrAus replied:
> The adapter would have to serialize/format all the field values individually as they come in before exposing them to the other framework.

That's kind of what I was thinking too. Capture values into a `serde_json::Value`-like thing. If we get to the point of designing the record API for maps and sequences then getting a chance to box an anonymous value might make that serialization a bit cheaper.

> We should ensure that the choice of whether or not to allocate here is left to the consumer of the trace data.

That seems fair. It's reasonable for `tokio-trace` to optimise its design around being the sole provider of events in an application, and work to keep that path as allocation free as possible. For integration points where paradigms meet we'll probably have to accept some cost.

So all that's to say I like the way this works.

hawkw replied:
> So all that's to say I like the way this works.

That's good to hear, thanks! :D

stjepang commented:
@hawkw That's a really elaborate answer, thanks! :)

> I was hoping that [this section](https://github.com/tokio-rs/tokio/blob/399192352ee76b42f182d37f8e87ef9a98bd2a12/tokio-trace/src/span.rs#L31-L74) in the `Span` docs would explain why spans might be re-entered. If that wasn't clear enough, is there anything that stands out to you that could be improved there?

Ah, I was looking at the readme and not into doc comments. The comments explain spans well.

> for example, if we want to know how long was spent in the loop overall, versus if we wanted to know how much time was spent in each individual iteration

That's a nice way of putting it, and now seems obvious in hindsight.

> Perhaps the docs should link to examples in the nursery repo?

I'd find that helpful.

Just two more questions:

* Closing spans sounds conceptually similar to closing channels. Dropping the last handle closes it, but you can also manually call `.close()`. I'm surprised that in `my_span.close(); my_span.enter(|| ...)` the span will be enered one more time and *then* get closed. Is there a reason why `.close()` doesn't close the span immediately?

* How does `enter(|| ...)` interact with async/await? It seems like `await fut` inside a span will not work because of the inner closure. However, if we introduce a magical `enter!(span, { ... })` macro, then awaiting a future inside the block could *just work*. Is my understanding correct?

hawkw commented:
@stjepang 
> * Closing spans sounds conceptually similar to closing channels. Dropping the last handle closes it, but you can also manually call `.close()`. I'm surprised that in `my_span.close(); my_span.enter(|| ...)` the span will be enered one more time and _then_ get closed. Is there a reason why `.close()` doesn't close the span immediately?

Yikes! This behaviour is an artifact of an earlier design, where span closure occurred as part of `Subscriber::exit` --- I thought I changed how `Span::close` behaved when that was removed, however, that doesn't appear to be the case! This definitely needs to be fixed, thanks for noticing that.

> * How does `enter(|| ...)` interact with async/await? It seems like `await fut` inside a span will not work because of the inner closure. However, if we introduce a magical `enter!(span, { ... })` macro, then awaiting a future inside the block could _just work_. Is my understanding correct?

I'll be honest that I haven't experimented too much with `tokio-trace`/async-await compatibility, but that does sound possible --- a macro may be necessary. Hopefully that can be an additive change.

carllerche commented:
We should set a max number of span fields. 32 is a good start. This ensures we can use stack arrays for passing fields around.

carllerche commented on tokio-trace/tokio-trace-core/src/metadata.rs:
```diff
@@ -0,0 +1,272 @@
+        module_path: Option<&'a str>,
+        file: Option<&'a str>,
+        line: Option<u32>,
+        field_names: &'static [&'static str],
```
`field_names` should not be exposed as `&[&str]`. We probably want an opaque type for this.

hawkw replied:
In theory, I agree. In practice, since we can't use `const fn`s yet, the opaque type would need to have another constructor macro, so it just feels like a bunch of additional complexity to me. I can make that change regardless.

hawkw replied:
(also, I think `FieldSet` is essentially the opaque type for this, but it also has a callsite ID. we can't construct that here since we need to create the metadata before the callsite)

KodrAus commented:
Having some arbitrary limit on the number of fields or the nested depth of complex objects is also, in my opinion, a defensible philosophical decision. It makes you stop and think about what it is you're actually collecting.

hawkw commented:
FWIW I believe the AppVeyor CI build for f2c8354fd975913d330927a497b74c690b09a46a failed due to an unrelated issue.
